### DISCUSSION TITLE
Module 'SCHAModules' has no attribute 'thermodynamic'

### DISCUSSION THREAD

**Discussion by szkudlarek-piotr** [SSCHAcode/python-sscha#194] [Q&A] [answered]
Dear SSCHA community, 
I am trying to generate an ensemble to canculate anharmonic phonons. Here is the code I am using:

```
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble
import numpy as np
np.random.seed(0)
dyn = CC.Phonons.Phonons("LiBaF3", nqirr=6)

temperature = 300
ensemble = sscha.Ensemble.Ensemble(dyn, temperature)
number_of_configurations = 100
ensemble.generate(number_of_configurations)
save_directory = "data"
population_id = 1
ensemble.save(save_directory, population_id)  
```
which was greatly influenced by the materials delivered at SSCHA school.. However, when I am running it, I am getting an error shown below:
**Traceback (most recent call last):
  File "/home/max/Desktop/LiBaF3_single/gen1.py", line 18, in <module>
    ensemble.save(save_directory, population_id)
  File "/home/max/.conda/lib/python3.10/site-packages/sscha/Ensemble.py", line 892, in save
    self.update_weights(self.dyn_0, self.current_T)
  File "/home/max/.conda/lib/python3.10/site-packages/sscha/Ensemble.py", line 1950, in update_weights
    old_a = SCHAModules.thermodynamic.w_to_a(w, self.T0)
AttributeError: module 'SCHAModules' has no attribute 'thermodynamic'**
Is there anything wrong with my code? Is there any update of the code that caused my problem?
Thank you in advance,
Piotr Szkudlarek

**Comment by diegomartinez2** [comment, marked as answer]
There was an update in some related libraries that broke part of our code. We solved this in the last patch. You can reinstall SSCHA or use the solution of #184

**Reply by mesonepigreco** [reply]
Just an addition: reinstall from the main branch of the GitHub repository. As of today, 14 February 2024, we still have to release the patch on pip.

**Reply by mesonepigreco** [reply]
Ok, now the patch has been published also on pip, in version 1.4.1 of both cellconstructor and python-sscha.
So you can update them with
```
pip install -U cellconstructor
pip install -U python-sscha
```

And then the code should work

**Reply by diegomartinez2** [reply]
If you get an error with the pip re-installation (something like "It is a distutils installed project and thus we cannot accurately determine..."), you have installed the package using ``python setup.py install``. 
In this case, either pull the master to its last version and reinstall it in the same way or download the source of the latest release and install it with python setup.py install.

If you want to rely on pip, you can force the reinstallation in the following way, but
<b>be aware that doing this could break your python environment</b>. It is better to create a new python environment and start there a fresh installation.

Anyway, if you really want to proceed, these are the commands

```
# Do not use this command if you do not know what you are doing
pip install --ignore-installed cellconstructor
pip install --ignore-installed python-sscha
```
This will overwrite the previous version without removing the previous version so be aware of the "trash" it leaves behind.

**Comment by szkudlarek-piotr** [comment]
Thank you for your help,
Now, when I am trying the same thing with the folloorwing code:
```
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble
import numpy as np
# Load the dynamical matrix
dyn = CC.Phonons.Phonons("start_sscha", nqirr=6).Symmetrize().ForcePositiveDefinite()
#[ apply here the needed changes to dyn ]
# Prepare the ensemble
temperature = 300
ensemble = sscha.Ensemble.Ensemble(dyn, temperature)
# Generate the ensemble
number_of_configurations = 100
ensemble.generate(number_of_configurations)
# Save the ensemble into a directory
save_directory = "data"
population_id = 1
ensemble.save(save_directory, population_id)

```
I am getting an error:
```
Traceback (most recent call last):
  File "/home/max/Desktop/single_LiBaF3/gen1.py", line 5, in <module>
    dyn = CC.Phonons.Phonons("start_sscha", nqirr=6).Symmetrize().ForcePositiveDefinite()
AttributeError: 'NoneType' object has no attribute 'ForcePositiveDefinite'
```
As far as I understand, it doesn't treat dyn as a dynamical matrix. Is it the fault of the data or my processing?

**Reply by diegomartinez2** [reply]
Maybe an error occurs during the creation of the object, causing the function to return None instead of the object.

Check if the class is not returning an instance with:
```
test = CC.Phonons.Phonons("start_sscha", nqirr=6)
if test is not None:
    test.ForcePositiveDefinite()
else:
    print("The function returned None.")

```

**Comment by szkudlarek-piotr** [comment]
Both .Symmetrize() and .ForcePositiveDefinite() transform instance of cellconstructor.Phonons.Phonons to NoneType.

**Comment by mesonepigreco** [comment]
> Both .Symmetrize() and .ForcePositiveDefinite() transform instance of
> cellconstructor.Phonons.Phonons to NoneType.

Yes, they should be called on the dyn, as they do not return the object
itself.
So you must call then after you created the dyn in subsequent instructions:

```python
dyn=CC.Phonons.Phonons(...)
dyn.Symmetrize()
dyn.ForcePositiveDefinite()
```

Symmetrize and ForcePositiveDefinite modify inplace the dyn object, without
returning anything.

**Reply by szkudlarek-piotr** [reply]
That solved my problem! Grazie mille!
